### PR TITLE
MSI build with PROFILE / ARP/ UI & update NSI

### PR DESCRIPTION
**Pull request by Delaire**
UPDATED:
Adding a first draft on building an MSI using Advanced Installer, can handle the properties PROFILE/ ARP/ UI.

The saving of the properties was moved outside of the aip project so that it can be move easily updated, it is located in the powershell script called InstallProperties.ps1

Examples:
msiexec.exe /i "NextDNS.msi" /qn PROFILE=newKey01 UI=0 ARP=0

Update service so we can read the ARP and UI values in the reg
Handle OnStart event to check to see we have Profile ID and UI=False to autostart service without having to use client app

Also adding nsis exe build update to handle ARP and UI DWORD from registry

### CODE REVIEW

File: msi/NextDNS.aip [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+  <COMPONENT cid="caphyon.advinst.msicomp.ProjectOptionsComponent">
+    <ROW Name="HiddenItems" Value="AppXProductDetailsComponent;AppXDependenciesComponent;AppXAppDetailsComponent;AppXVisualAssetsComponent;AppXCapabilitiesComponent;AppXAppDeclarationsComponent;AppXUriRulesComponent;SccmComponent;ActSyncAppComponent;CPLAppletComponent;AutorunComponent;GameUxComponent;SilverlightSlnComponent;FixupComponent;MsiXDiffComponent;MsixManifestEditorComponent"/>
+  </COMPONENT>
+  <COMPONENT cid="caphyon.advinst.msicomp.MsiPropsComponent">
+    <ROW Property="AI_BITMAP_DISPLAY_MODE" Value="0"/>
+    <ROW Property="AI_CURRENT_YEAR" Value="2022" ValueLocId="-"/>
+    <ROW Property="AI_PACKAGING_TOOL" Value="Advanced Installer 19.5 build 36301275" ValueLocId="-"/>
+    <ROW Property="ALLUSERS" Value="1"/>
+    <ROW Property="ARPCOMMENTS" Value="This installer database contains the logic and data required to install [|ProductName]." ValueLocId="*"/>
+    <ROW Property="ARPCONTACT" Value="team@nextdns.io"/>
```
Comment (rs) [inline msi/NextDNS.aip:12]: Are contact and phone properties required?

File: msi/install_examples.txt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+msiexec.exe /i "NextDNS.msi" /qn PROFILE=newKey01 UI=no ARP=no
```
Comment (rs) [inline msi/install_examples.txt:1]: Content of this file should be moved to the README

**Comment by rs** [comment]
Can you please add the building of the msi to the main build script?

File: msi/NextDNS.aip (hunk 1 above)
Comment (Delaire) [reply msi/NextDNS.aip:12]: No it is not, ill remove.
image

File: msi/NextDNS.aip [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  <COMPONENT cid="caphyon.advinst.msicomp.ProjectOptionsComponent">
+    <ROW Name="HiddenItems" Value="AppXProductDetailsComponent;AppXDependenciesComponent;AppXAppDetailsComponent;AppXVisualAssetsComponent;AppXCapabilitiesComponent;AppXAppDeclarationsComponent;AppXUriRulesComponent;SccmComponent;ActSyncAppComponent;CPLAppletComponent;AutorunComponent;GameUxComponent;SilverlightSlnComponent;FixupComponent;MsiXDiffComponent;MsixManifestEditorComponent"/>
+  </COMPONENT>
+  <COMPONENT cid="caphyon.advinst.msicomp.MsiPropsComponent">
+    <ROW Property="AI_BITMAP_DISPLAY_MODE" Value="0"/>
+    <ROW Property="AI_CURRENT_YEAR" Value="2022" ValueLocId="-"/>
+    <ROW Property="AI_PACKAGING_TOOL" Value="Advanced Installer 19.5 build 36301275" ValueLocId="-"/>
+    <ROW Property="ALLUSERS" Value="1"/>
+    <ROW Property="ARPCOMMENTS" Value="This installer database contains the logic and data required to install [|ProductName]." ValueLocId="*"/>
+    <ROW Property="ARPHELPLINK" Value="https://nextdns.io/"/>
```
Comment (rs) [inline msi/NextDNS.aip:12]: This one is https://help.nextdns.io

**Comment by Delaire** [comment]
> Can you please add the building of the msi to the main build script?

I have update the build.bat file with advanced installer cli to build the msi

File: build.bat [hunk 4]
```diff
@@ -33,6 +34,8 @@ goto :init
     set "__BAT_NAME=%~nx0"
 
     set "OptNoSign="
+    set "OptBuildMsi="
+	set "PathExe="
```
Comment (rs) [inline build.bat:38]: Mixed identation

File: msi/PowerShellScriptInline/InstallProperties.ps1 [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	)
+	{
+		New-ItemProperty -Path $registryPath -Name "UI" -Value 0 -PropertyType DWord -Force | Out-Null
+		New-ItemProperty -Path $registryPath -Name "ARP" -Value 0 -PropertyType DWord -Force | Out-Null
+	}
+	
+	# If skip contains UI
+	IF($propSkipValue.ToUpper() -contains "UI")
+	{
+		New-ItemProperty -Path $registryPath -Name "UI" -Value 0 -PropertyType DWord -Force | Out-Null
```
Comment (rs) [inline msi/PowerShellScriptInline/InstallProperties.ps1:37]: If only UI, make sure ARP is removed

File: msi/PowerShellScriptInline/InstallProperties.ps1 [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+	}
+
+	# If skip contains UI
+	IF($propSkipValue.ToUpper() -contains "ARP")
+	{
+		New-ItemProperty -Path $registryPath -Name "ARP" -Value 0 -PropertyType DWord -Force | Out-Null
```
Comment (rs) [inline msi/PowerShellScriptInline/InstallProperties.ps1:43]: If only ARP, make sure UI is removed

File: msi/PowerShellScriptInline/InstallProperties.ps1 [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+IF(![string]::IsNullOrEmpty($propProfileValue))
+{
+	New-ItemProperty -Path $registryPath -Name "Profile" -Value $propProfileValue -PropertyType "String" -Force | Out-Null
+}
+
+# SKIP
+IF(![string]::IsNullOrEmpty($propSkipValue))
+{
```
Comment (rs) [inline msi/PowerShellScriptInline/InstallProperties.ps1:25]: Can we detect if SKIP is provided with no value (as opposed to no provided at all)? If SKIP is provided with no value, it should be a value to reset all skips (remove the regs).

File: README.md [hunk 8]
```diff
@@ -20,5 +20,27 @@
...
+  * SKIP=UI allows you to install the application without the UI
+  * SKIP=ARP allows you to hide the list of the app in Add/Remove Program
+
+#### Examples
+
+Using SKIP:
+
+```msiexec.exe /i "NextDNS.msi" /qn PROFILE=newKey01 SKIP=UI/ARP```
+
+```msiexec.exe /i "NextDNS.msi" /qn PROFILE=newKey01 SKIP=UI_AND_ARP```
```
Comment (rs) [inline README.md:43]: Please support `UI,ARP` and only that (no `_AND_` or `/`).

File: msi/PowerShellScriptInline/InstallProperties.ps1 [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+	# If skip contains UI
+	IF($propSkipValue.ToUpper() -contains "UI_AND_ARP" -or $propSkipValue.ToUpper() -contains "ARP_AND_UI"	-or $propSkipValue.ToUpper() -contains "UI/ARP" -or $propSkipValue.ToUpper() -contains "ARP/UI"
```
Comment (rs) [inline msi/PowerShellScriptInline/InstallProperties.ps1:27]: What about splitting the string on `,` and then test for the presence of each supported flag?

**Review by rs — CHANGES_REQUESTED** [review]

File: README.md [hunk 10]
```diff
@@ -8,7 +8,7 @@
   * In Visual Studio, install the following components
     * .NET desktop development
     * Desktop development with C++
-    * Universal Windows Platform development 
+    * Universal Windows Platform development
 * Install [Windows SDK](https://dev.windows.com/en-us/downloads/windows-10-sdk)
 * Install [Windows WDK](https://docs.microsoft.com/en-us/windows-hardware/drivers/download-the-wdk)
 * Install [NSIS](https://nsis.sourceforge.io/Download)
```
Comment (rs) [inline README.md:14]: Don't we need to install something to build the MSI?

**Comment by Delaire** [comment]
Ill redo a full check of the nsis install / uninstall , with and with out reg keys before the end of the week so that we are ok on what has been done

**Comment by Delaire** [comment]
I have tested the following install options for exe and msi, do you think I need to test something else?
image

**Pull request merged**